### ISSUE TITLE
Creating an XMLHttpRequest to pull file and send it to S3

### ISSUE THREAD

**Issue by vinhdang84** [vinhdang84/campaign-data-downloader#3]
I have a file that is downloaded. I want to upload this file to S3. 
* I want to upload the file to make sure it looks good. 
* I want to optimize the file in S3 and not within the browser. 
* If in the future, I want to change the logic in how the file is parsed, I want to make the change in S3 and not in the extension which would require a new file download. The extension should be as simple as possible.

**Comment by vinhdang84** [comment]
Example of grabbing the file and uploading it 

https://stackoverflow.com/questions/4093722/upload-a-file-in-a-google-chrome-extension

**Comment by vinhdang84** [comment]
https://developer.chrome.com/extensions/xhr


https://developers.google.com/web/updates/2015/03/introduction-to-fetch

**Comment by vinhdang84** [comment]
Reading the below, his approach to the problem is to: 
1. Pull the URL for the image which will create a message with this URL in the content script. 
2. In the background script, the script will listen for an event for this URL message. I'm not familiar with web worker yet, so I need to look into what this means. "var worker = new Worker('worker.js');" But it seems it posts another message using webworker. 
3. 
*Webworker will grab the message data from the background script, GET the url (HttpProtocol) expecting that it is coming from an array. 
*When the state of the object xhr is 200 OK it will run the nextStep function utilizing the pulled data from the URL (xhr.response). 
* For convenience, he uses append so that there is an array with the URL data, with server method: upload, and file: array buffer. He opens the php form on the remote server, and sends this array to the php form. 

> The following approach was used:
> 
> File downloads: Get the src property of an <img> element, for example.
> Fetch the file from the Cache - use XMLHttpRequest from the background page.
> Use a Web Worker in the background page to handle the upload.
> Side note, to take the checksum of the image, Crypto-JS: MD5 can be used. Example (where xhr is the XMLHttpRequest object with responseType set to arraybuffer, see Worker demo):
> 
> var md5sum = Crypto.MD5( new Uint8Array(xhr.response) );
> Full example
> Content script
> // Example: Grab the first <img> from the document if it exists.
> var img = document.images[0];
> if (img) {
>     // Send the target of the image:
>     chrome.runtime.sendMessage({method: 'postUrl', url: img.src});
> }
> Background script (with Worker)
> chrome.runtime.onMessage.addListener(function(request) {
>     if (request.method == 'postUrl') {
>         var worker = new Worker('worker.js');
>         worker.postMessage(request.url);
>     }
> });
> Web Worker
> // Define the FormData object for the Web worker:
> importScripts('xhr2-FormData.js')
> 
> // Note: In a Web worker, the global object is called "self" instead of "window"
> self.onmessage = function(event) {
>     var resourceUrl = event.data;   // From the background page
>     var xhr = new XMLHttpRequest();
>     xhr.open('GET', resourceUrl, true);
> 
>     // Response type arraybuffer - XMLHttpRequest 2
>     xhr.responseType = 'arraybuffer';
>     xhr.onload = function(e) {
>         if (xhr.status == 200) {
>             nextStep(xhr.response);
>         }
>     };
>     xhr.send();
> };
> function nextStep(arrayBuffer) {
>     var xhr = new XMLHttpRequest();
>     // Using FormData polyfill for Web workers!
>     var fd = new FormData();
>     fd.append('server-method', 'upload');
> 
>     // The native FormData.append method ONLY takes Blobs, Files or strings
>     // The FormData for Web workers polyfill can also deal with array buffers
>     fd.append('file', arrayBuffer);
> 
>     xhr.open('POST', 'http://YOUR.DOMAIN.HERE/posturl.php', true);
> 
>     // Transmit the form to the server
>     xhr.send(fd);
> };

**Comment by vinhdang84** [comment]
Compared to what is written in the stackoverflow example, I have a handful of differences. 
- I already know the URL so I don't need a content script to grab the URL on a page. 
- Don't need an event listener to listen for the URL message in background.js
- Don't need to post the URL in background.js as a message to another script. 
- My current script download the file locally which requires main.js to call the function in my background script called download.js. Both the function in download.js isn't needed along with main.js to call for that background script's function to execute. 
- I still want the file to by manually done on button press so I would still need my main.html to have a button and a listener on main.js to execute XMLHttpRequest(). 

Questions: 
Step 10-13 in the below will depend on S3. I will need to create another issue which is how do I upload a file to s3. I will need to look at permissions in IAM, what the data needs to look like on sending the message (it may not require form data), and the method of sending XMLHttpRequest to s3. 

Do I need background script to call web worker and post message of URL? My belief and first test will be running this all without a background script. The background script is needed in his example because it will likely execute on page load where an image appears. Mine will run on button click within the extension's popup & has an absolute path without the need to open a page. 



What is needed: 
 - main.html
 - main.js

1. Keep button on main.html with event click and execution of main.js in background.
2. Keep event listener for button click on main.js
3. Remove function to run background page within main.js
4. Main.js will create a variable (xhr) for new XMLHttpeRequest(). 
5. It will GET the URL 
6. Find best responseType for a file online and set the responseType. 
7. Set xhr.onload to a function which will upon finding that status == 200, to run a function called nextStep with the response (xhr.response) of the xhr.open. 
8. Create function called nextStep which contains a variable called "csvFile" that will be filled in with the response from xhr.open
9. Create variable xhr which is set to new XMLHttpRequest()
10. Create variable  fd set to new FormData()
11. Append to this variable server-method: upload and file: csvFile
12. xhr.open which will send a message to POST to the remote server with option of 'true'. (read into why option true is needed) 
13. Send file using xhr.send(fd)

**Comment by vinhdang84** [comment]
Note: May need to change manifest.json permission to get an origin header which will allow for CORS 
https://stackoverflow.com/questions/24586254/corss3browser-cachechrome-extension

**Comment by vinhdang84** [comment]
https://support.google.com/chrome/a/answer/2714278?hl=en

https://stackoverflow.com/questions/9421933/cross-origin-xmlhttprequest-in-chrome-extensions

The question stated here is how to run XMLHttpRequest in chrome dev. It appears you need to pack the extension.
"Two things; you need to make sure you are making a packaged app/extension and not a hosted one. Cross origin requests will not work with hosted apps. Assuming you got that part pinned down, you may want to try to put the following into your permissions: http://*/ . That's the only one I have for one of my packaged apps, and it does cross origin stuff without any problems.


https://developers.google.com/chrome/web-store/articles/apps_vs_extensions?hl=zh-CN
